lte/gateway/c/session_manager/SessionState.cpp

```diff
@@ -1351,35 +1354,66 @@ bool SessionState::receive_monitor(
 
   if (it == monitor_map_.end()) {
     // new credit
-    return init_new_monitor(update, update_criteria);
+    return init_new_monitor(update, session_uc);
   }
-  auto credit_uc = get_monitor_uc(mkey, update_criteria);
+  auto credit_uc = get_monitor_uc(mkey, session_uc);
   if (!update.success()) {
     it->second->credit.mark_failure(update.result_code(), credit_uc);
     return false;
   }
 
-  MLOG(MINFO) << session_id_ << " Received monitor credit for " << mkey;
-  const auto& gsu = update.credit().granted_units();
-  it->second->credit.receive_credit(gsu, credit_uc);
+  if (update.credit().action() == UsageMonitoringCredit::FORCE){
+    MLOG(MINFO) << "UsageMonitoringCredit::FORCE "
+                   "(`AVP: Usage-Monitoring-Report`) instruction "
+                   "not implemented. Will just continue for monitor  " << mkey;
+  }
+
+  if (update.credit().action() == UsageMonitoringCredit::DISABLE) {
+    // Disable mpnitor, no matter if we still have credit
+    MLOG(MINFO) << "Received a forced Disabled action for monitor " << mkey
+                << ". Will remove monitor after update is sent ";
+    // seting last update will deleted monitor after the update is sent.
+    it->second->credit.set_last_update(true, *credit_uc);
+
+  } else {
+    MLOG(MINFO) << session_id_ << " Received monitor credit for " << mkey;
+    const auto& gsu = update.credit().granted_units();
+    it->second->credit.receive_credit(gsu, credit_uc);
+  }
+  
   return true;
 }
 
 void SessionState::apply_monitor_updates(
-    const std::string& key, SessionCreditUpdateCriteria& update) {
+    const std::string& key, SessionStateUpdateCriteria& session_uc,
+    SessionCreditUpdateCriteria& credit_uc) {
   auto it = monitor_map_.find(key);
   if (it == monitor_map_.end()) {
     return;
   }
 
-  auto& credit = it->second->credit;
+  // Actual deletion of monitor
+  if (credit_uc.deleted) {
+    if (it->second->level == MonitoringLevel::SESSION_LEVEL){
+      // session level change
+      MLOG(MINFO) << "Removing Session Level monitor " << key;
+      session_uc.is_session_level_key_updated = true;
+      session_uc.updated_session_level_key = "";
+    }
+    MLOG(MINFO) << "Erasing monitor " << key;
```

Comment:
I'm starting to think we should just add a SessionState logging function that just adds the session_id automatically... but anyhow, pls add sessionID :D